### PR TITLE
readUBC: loading either 2D or 3D mesh

### PR DESCRIPTION
**Pull request by thast**
- discretize.TensorMesh.readUBC is now a wrapper for loading either 2D or 3D mesh
-  discretize.TensorMesh.readUBC can take care of recognizing if the file is a 2D or 3D UBC mesh, or it can be an input given by the user (arg: *meshdim*)
- the original functions are now *hidden* class function under discretize.TensorMesh._readUBC_3DMesh and discretize.TensorMesh._readUBC_2DMesh
- comments in UBC files are assumed to be marked with "!" in front

### CODE REVIEW

**Comment by thast** [comment]
I tested it on one of my mesh and it worked. Is there anything to do with regards to testing for discretize?

**Comment by lheagy** [comment]
@thast: you can add a test similar to https://github.com/simpeg/discretize/blob/master/tests/base/test_tensor_io.py#L50. A small ubs mesh file is included in the test directory, loaded in, and some sanity-checks run

**Comment by lheagy** [comment]
@thast: would you mind updating this branch so it is inline with the latest master?

**Comment by thast** [comment]
job 267 1 - simpeg discretize - travis ci 2017-05-07 13-02-10

**Comment by lheagy** [comment]
@thast: I have restarted the tests. If it was just a link failure, then hopefully they should pass.

File: discretize/MeshIO.py [hunk 1]
```diff
@@ -44,6 +44,63 @@ def readCellLine(line):
...
+    @classmethod
+    def readUBC_DC2DMesh(TensorMesh,fileName):
+        """
+        Read UBC GIF 2DTensor mesh and generate 2D Tensor mesh in simpeg
+
+        :param string fileName: path to the UBC GIF mesh file
+        :rtype: TensorMesh
+        :return: SimPEG TensorMesh 2D object
+
+        Created on Thu Nov 12 13:14:10 2015
```
Comment (lheagy) [inline discretize/MeshIO.py]: I would suggest not necessarily tracking this - all of this info is in the git log

File: discretize/MeshIO.py [hunk 2]
```diff
@@ -44,6 +44,63 @@ def readCellLine(line):
...
         return tensMsh
+    
+    @classmethod
+    def readUBC_DC2DMesh(TensorMesh,fileName):
+        """
+        Read UBC GIF 2DTensor mesh and generate 2D Tensor mesh in simpeg
+
+        :param string fileName: path to the UBC GIF mesh file
+        :rtype: TensorMesh
+        :return: SimPEG TensorMesh 2D object
```
Comment (lheagy) [inline discretize/MeshIO.py]: here, this is a `discretize TensorMesh` now :)

File: discretize/MeshIO.py [hunk 3]
```diff
@@ -44,6 +44,63 @@ def readCellLine(line):
...
+        @author: dominiquef
+
+        """
+
+        fopen = open(fileName, 'r')
+
+        # Read down the file and unpack dx vector
+        def unpackdx(fid, nrows):
+            for ii in range(nrows):
+
```
Comment (lheagy) [inline discretize/MeshIO.py]: would you mind opening up this file in a text editor with a pep8 linter? there is a bit of whitespace cleanup that can be done (eg within the function there should be at most one line between each code statement). To get up and running with a sublime pep8 linter, please see https://github.com/SublimeLinter/SublimeLinter-pep8

File: discretize/MeshIO.py [hunk 4]
```diff
@@ -44,6 +44,63 @@ def readCellLine(line):
...
+                else:
+                    xvec = np.hstack((xvec, np.ones(int(var[1])) * (var[0] - xend) / int(var[1])))
+                    xend = var[0]
+
+            return x0, xvec
+
+        # Start with dx block
+        # First line specifies the number of rows for x-cells
+        line = fopen.readline()
+        nl = np.array(line.split(), dtype=float)
```
Comment (lheagy) [inline discretize/MeshIO.py]: This is causing the failure (https://travis-ci.org/simpeg/discretize/jobs/229743764#L6327). This should be `dtype=int` ?? 
image

File: discretize/MeshIO.py (hunk 2 above)
Comment (thast) [inline discretize/MeshIO.py]: done

File: discretize/MeshIO.py (hunk 1 above)
Comment (thast) [inline discretize/MeshIO.py]: done

File: discretize/MeshIO.py (hunk 3 above)
Comment (thast) [inline discretize/MeshIO.py]: I did. Thanks. Should have take care of most of it.

File: discretize/MeshIO.py (hunk 4 above)
Comment (thast) [inline discretize/MeshIO.py]: yep, that was it.

**Comment by lheagy** [comment]
@thast: could you please change the base of this pr so that it goes on to `dev` instead of `master`? Thanks!
image

**Comment by thast** [comment]
Not sure now why one Travis Pass and not the other one (it does not seem to come from a file I change). any idea?
job 309 1 - simpeg discretize - travis ci 2017-05-29 13-30-54

**Pull request merged**